firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp

```diff
@@ -45,46 +44,49 @@ bool validate_start_of_frame(const Frame &new_frame) {
  */
 FrameInputStatus validate_frame(const Frame &new_frame) {
   static const uint8_t mask_start_of_frame = 0x80;
-  if (new_frame[0] == 0x01 && (new_frame[1] & mask_start_of_frame) == mask_start_of_frame) {
-    if (((new_frame[0] + new_frame[1] + new_frame[2] + new_frame[3]) % (UINT8_MAX + 1)) ==
-        new_frame[4]) {
-      return FrameInputStatus::output_ready;
-    }
+  if (!validate_frame_header(new_frame, mask_start_of_frame)) {
+    return FrameInputStatus::invalid_header;
   }
 
-  return FrameInputStatus::checksum_failed;
+  if (!validate_frame_checksum(new_frame)) {
+    return FrameInputStatus::invalid_checksum;
+  }
+  return FrameInputStatus::output_ready;
 }
 
 FrameInputStatus FrameReceiver::update_frame_buffer(uint8_t new_byte) {
+  static const uint8_t mask_start_of_packet = 0x81;
   Frame frame_buffer;
 
-  if (frame_buf_.input(new_byte) == FrameBufferStatus::waiting) {
+  if (frame_buf_.input(new_byte) == BufferInputStatus::ok) {
     return FrameInputStatus::ok;
   }
 
```

Comment:
No! you're right, I have now added a switch block to deal with all 3 cases.

Reply:
It would be helpful to add a code comment explaining why we want to break after `BufferInputStatus::full`, and how this interacts with the behavior of `FrameBuffer`.